### PR TITLE
Add Linux symlink temp directory cleanup.

### PR DESCRIPTION
**Pull request by mikem8361**
Issue: https://github.com/microsoft/clrmd/issues/698

Added 3.1.0 runtime to global.json so the tests can run.

### CODE REVIEW

File: src/Microsoft.Diagnostics.Runtime/src/DataTargets/DataTarget.cs [hunk 1]
```diff
@@ -195,6 +199,28 @@ private ImmutableArray<ClrInfo> GetOrCreateClrVersions()
...
+        /// Cleans up the temporary directory and symlinks for the Linux DAC workaround.
+        /// </summary>
+        private void CleanupTempDirectory()
+        {
+            string? tempDirectory = Interlocked.Exchange(ref _tempDirectory, null);
+            if (tempDirectory != null)
+            {
+                try
+                {
+                    foreach (string file in Directory.EnumerateFiles(tempDirectory))
```
Comment (NextTurn) [inline src/Microsoft.Diagnostics.Runtime/src/DataTargets/DataTarget.cs:212]:
```suggestion
                    Directory.Delete(tempDirectory, recursive: true);
```

**Comment by mikem8361** [comment]
@leculver, @NextTurn what do you guys think about adding a public DataTarget "TempDirectory" property that lazily creates it like https://github.com/microsoft/clrmd/blob/12da3fcbf17df12257a23de274108a6fd6698baa/src/Microsoft.Diagnostics.Runtime/src/DataTargets/DataTarget.cs#L167

dotnet-dump, SOS, etc. creates a similar per-process temp directory and it would be nice to share it and the cleanup.

File: src/Microsoft.Diagnostics.Runtime/src/DataTargets/DataTarget.cs [hunk 2]
```diff
@@ -164,10 +168,10 @@ private ImmutableArray<ClrInfo> GetOrCreateClrVersions()
                     {
                         // Works around issue https://github.com/dotnet/coreclr/issues/20205
                         int processId = Process.GetCurrentProcess().Id;
-                        string tempDirectory = Path.Combine(Path.GetTempPath(), "clrmd" + processId);
-                        Directory.CreateDirectory(tempDirectory);
+                        _tempDirectory = Path.Combine(Path.GetTempPath(), "clrmd" + processId);
+                        Directory.CreateDirectory(_tempDirectory);
 
-                        string symlink = Path.Combine(tempDirectory, dacFileName);
+                        string symlink = Path.Combine(_tempDirectory, dacFileName);
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/DataTargets/DataTarget.cs:174]: This design creates a subtle race between two ClrMD processes inspecting the same target process.

If Process A is in `CleanupTmpDirectory` in or just before `Directory.Delete` and Process B has just returned from `LinuxFunctions.symlink` then Process A may delete our symlink before we have a chance to open it, leading to an odd exception in Process B when it can't find the dac symlink we created.

File: src/Microsoft.Diagnostics.Runtime/src/DataTargets/DataTarget.cs [hunk 3]
```diff
@@ -195,6 +199,24 @@ private ImmutableArray<ClrInfo> GetOrCreateClrVersions()
             return _clrs;
         }
 
+        /// <summary>
+        /// Cleans up the temporary directory and symlinks for the Linux DAC workaround.
+        /// </summary>
+        private void CleanupTempDirectory()
+        {
+            string? tempDirectory = Interlocked.Exchange(ref _tempDirectory, null);
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/DataTargets/DataTarget.cs:207]: Why do we need `Interlocked.Exchange` here?  The user shouldn't call Dispose twice and it won't hurt if they do.  Using `Interlocked.Exchange` doesn't _hurt_ here but it makes the code a little harder to understand, as in someone walking up to the codebase might think this is part of some multithreaded locking scheme when it's not.

I'd normally recommend replacing this line with `if (_tempDirectory != null && Directory.Exists(_tempDirectory)` and never assigning `_tempDirectory` back to null.  It's ok if the user mistakenly calls Dispose, even from multiple threads, since the catch should protect us below.  (However, as mentioned above there's still a race in this overall design, so this is general feedback that I don't expect you to address before we figure out a design that doesn't race.)

File: global.json [hunk 4]
```diff
@@ -1,6 +1,11 @@
 {
   "tools": {
```
Comment (leculver) [inline global.json:2]: What is the purpose of the change to global.json?  Is that expected in this review?

Comment (leculver) [reply global.json:2]: Oh I missed that comment.

**Comment by leculver** [comment]
> what do you guys think about adding a public DataTarget "TempDirectory" property that lazily creates it like
> https://github.com/microsoft/clrmd/blob/12da3fcbf17df12257a23de274108a6fd6698baa/src/Microsoft.Diagnostics.Runtime/src/DataTargets/DataTarget.cs#L167
> 
> dotnet-dump, SOS, etc. creates a similar per-process temp directory and it would be nice to share it and the cleanup.

I worry that we will create hard to track races between tools creating and cleaning up files within the same directory.  I wonder if we should go in a different direction.  Instead of creating a directory for symlinks, it might make more sense to not create a directory at all and instead do something like:

    _symlink = Path.GetTempFileName();
    if (LinuxFunctions.symlink(dacLocation, _symlink) == 0)
        dacLocation = symlink;

Then deleting our `_symlink` temp file in `Dispose`.

File: src/Microsoft.Diagnostics.Runtime/src/DataTargets/DataTarget.cs (hunk 2 above)
Comment (NextTurn) [reply src/Microsoft.Diagnostics.Runtime/src/DataTargets/DataTarget.cs:174]: Aha? The temporary directory has the ClrMD process' ID in its name.

**Comment by mikem8361** [comment]
I liked your last suggestion and push a new version.  It is a lot simpler.

File: src/Microsoft.Diagnostics.Runtime/src/DataTargets/DataTarget.cs (hunk 2 above)
Comment (leculver) [reply src/Microsoft.Diagnostics.Runtime/src/DataTargets/DataTarget.cs]: Ah you are right, I didn't notice it's the current process and not the target process.  What if two different threads in the same process create a DataTarget and clean it up?  I'm not sure the behavior of `LinuxFunctions.symlink` will be if two threads race to call that at the same time, will that fail and we won't set the dacLocation?  I think that same cleanup issues happens if two threads in the same process races.

File: src/Microsoft.Diagnostics.Runtime/src/DataTargets/DataTarget.cs [hunk 5]
```diff
@@ -195,6 +204,13 @@ private ImmutableArray<ClrInfo> GetOrCreateClrVersions()
             return _clrs;
         }
 
+        /// <summary>
+        /// Cleans up the temporary directory and symlinks for the Linux DAC workaround.
+        /// </summary>
+        private void CleanupTempDirectory()
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/DataTargets/DataTarget.cs:210]: We can remove this now right?

Comment (mikem8361) [reply src/Microsoft.Diagnostics.Runtime/src/DataTargets/DataTarget.cs:210]: Oops, I forgot to delete it.

**Comment by mikem8361** [comment]
I think with the new code, if two different threads enumerate the ClrInfos on the same DataTarget instance there could be a race so maybe I should put a lock around the creating the symlink:

```
              lock()
              {
                    if (_symlink == null)
                   {
                         _symlink = Path.GetTempFileName();
                        if (LinuxFunctions.symlink(dacLocation, _symlink) == 0)
                        {
                            dacLocation = _symlink;
                        }
                   }
              }
```

**Comment by leculver** [comment]
Yep good point on the race.

You could also use Interlocked.CompareExchange to avoid a lock here, just make sure the symlink is fully created before publishing the value, and delete the temp file for the thread that lost the race.

Is there any way to load two different .Net Core runtimes into a single Linux process?  For example, two runtimes living in parallel directories?  We don't have to address that problem today, but I'd just point out that the current design only works in a non SxS scenario.  It's a problem for a different day if this is possible.

**Comment by mikem8361** [comment]
Like you said before interlocked is harder to follow and I don't think a lock could cause any problems/deadlocks. I'm going to add a lock around the code unless you object.

Multiple .NET Core processes side-by-side isn't supported but actually might work.  The old hosting APIs (coreclr_initialize(), etc.) don't prevent it (the new ones may). It hasn't been tested in a long time.

We can have a desktop CLR and .NET Core runtime in the same process but that doesn't apply here.

**Comment by mikem8361** [comment]
One thing I didn't think about is that `lock(this)` isn't a great practice because outside code can lock the DataTarget instance might cause problems/deadlocks.  So adding a private object instance is the usual practice if there isn't already some private. Seems overkill to just allocate another object instance for a lock, but I might be overthinking this.

**Comment by leculver** [comment]
No worries, and definitely don't overthink it.  Just add a new `private readonly object _sync = new object();` to DataTarget.  I can always move it to interlocked later if need be.

**Comment by NextTurn** [comment]
@leculver @mikem8361 

(I'm cleaning up my ancient email backlog.) This doesn't actually work. `Path.GetTempFileName` will create the file and `symlink` will fail with `EEXIST`.

Besides, the synchronization looks odd... If there is a race between two threads the temp file will be created twice?
